---
title: ""
output:
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
setwd("/home/ciganche/Documents/Faks/II/rna_dgea")
library(biomaRt)
library(data.table)
library(tidyverse)
library(edgeR)
library(org.Hs.eg.db)
library(DOSE)
library(clusterProfiler)
library(pathview)
library(MotifDb)
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
library(igraph)
```
### Diffenrential gene expression analysis of a prostate cancer toy RNA-seq dataset using the state of the art methods for DE, enrichment, annotation and motif analysis. 
<br />
<br />

Load the prostate cancer dataset, generate using the Recount package.
```{r}
load("prostate_cancer.RData")
```
<br />
<br />

Extract protein coding genes:
```{r}
ensembl = useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
query1 = getBM(attributes=c("ensembl_gene_id", "gene_biotype"),
                filters=c("ensembl_gene_id"), 
                values=list(c(r_anno_df$gene_id)),
                mart = ensembl)
query1_protein_coding = query1[which(query1$gene_biotype=="protein_coding"),]

r_anno_df = subset(r_anno_df, r_anno_df$gene_id %in% query1_protein_coding$ensembl_gene_id)
raw_counts_df = subset(raw_counts_df, rownames(raw_counts_df) %in% query1_protein_coding$ensembl_gene_id)

cat(sprintf("Before selecting protein coding: %s \n After selecting protein coding: %s", nrow(query1), nrow(query1_protein_coding)))
```
<br />
<br />

Filter to 20 counts at least in 1 tumor and 1 normal, then normalize the counts:
```{r}
c_anno_case = subset(c_anno_df, c_anno_df$condition=="Case")
c_anno_control = subset(c_anno_df, c_anno_df$condition=="Control")

df_raw_case = raw_counts_df[,c(c_anno_case$sample)] 
over_20_case = apply(df_raw_case, 1, function(x) any(x > 20))

df_raw_control = raw_counts_df[,c(c_anno_control$sample)] 
over_20_control = apply(df_raw_control, 1, function(x) any(x > 20))

boolean = over_20_case & over_20_control
filtered_raw_df  = cbind(df_raw_case, df_raw_control, boolean)
c_anno_df = rbind(c_anno_case, c_anno_control)

filtered_raw_df = subset(filtered_raw_df, filtered_raw_df$boolean)
filtered_raw_df = filtered_raw_df[,-101]
rownames(r_anno_df) = r_anno_df$gene_id
filtered_anno_df = r_anno_df[rownames(filtered_raw_df),]

edge_dge = DGEList(counts=filtered_raw_df,group=c_anno_df$condition,samples=c_anno_df,genes=filtered_anno_df) 

edge_n = calcNormFactors(edge_dge,method="TMM")
cpm_table = as.data.frame(round(cpm(edge_n),2))

cat(sprintf("Before filtering: %s \n After filtering: %s", nrow(raw_counts_df), nrow(filtered_raw_df)))
```
<br />
<br />

Before normalization:
```{r}
long_counts_df = gather(filtered_raw_df, key = "sample", value = "read_number")
ggplot(data=long_counts_df,aes(sample,read_number+1)) +
   geom_boxplot(fill="orange",alpha=0.7)+
   theme_minimal() +
   scale_x_discrete(labels=NULL) +
   scale_y_log10()
```
<br />
<br />

After normalization - medians are aligned pretty well:
```{r}
long_cpm_df = gather(cpm_table, key = "sample", value = "CPM") 
ggplot(data=long_cpm_df,aes(sample,CPM+1)) +
   geom_boxplot(fill="green",alpha=0.7)+
   theme_minimal() +
   scale_x_discrete(labels=NULL) +
   scale_y_log10()
```
<br />
<br />

EdgeR analysis - thresholds for corrected p-val of 0.01 and fold change of 1.5 were used.  
```{r}
design = model.matrix(~0+group, data=edge_dge$samples)
colnames(design) = levels(edge_dge$samples$group)
rownames(design) = edge_dge$samples$sample

edge_d = estimateDisp(edge_n,design)
edge_f = glmQLFit(edge_d,design) 

contro = makeContrasts("Case-Control",levels=design) 

edgeRglmQLF = function(mat=edge_f,contro,cpm_mat=edge_n,label="Cancer vs Control",sig_thr=1,sig_col="log2_CPM",fc_thr=1.5,pval_col="p_val",pval_thr=0.01,names=FALSE)
{
   degs = glmQLFTest(edge_f,contrast=contro)$table[,-3]
   colnames(degs) = c("log2_FC","log2_CPM","p_val")
   a_levels = rownames(contro)[which(contro!=0)]
   a_samples = which(cpm_mat$samples$group%in%a_levels)
   cpm_sele = cpm(cpm_mat,log=T)[,a_samples]
   degs$log2_CPM = apply(cpm_sele,1,function(x) mean(x))
   degs$p_adj = p.adjust(degs$p_val, method ="BH")
   degs$class = "="
   degs[which(degs[,sig_col]>=sig_thr & degs$log2_FC>=fc_thr & degs[,pval_col]<=pval_thr),"class"] = "+"
   degs[which(degs[,sig_col]>=sig_thr & degs$log2_FC<=(-fc_thr) & degs[,pval_col]<=pval_thr),"class"] = "-"
   degs$class = as.factor(degs$class)
   degs$comp = label
   degs$id = rownames(degs)
   degs = degs[,c("id","comp","log2_FC","log2_CPM","p_val","p_adj","class")]
   if(names=="TRUE"){
      newnames = paste(label,colnames(degs),sep="_")
      colnames(degs) = newnames
   }
   return(degs)
}
DEGs = edgeRglmQLF(mat=edge_f, cpm_mat=edge_n, contro=contro, sig_thr=1, sig_col="log2_CPM", fc_thr=1.5, pval_thr=0.01, pval_col="p_adj",names=F)

summary(DEGs$class)
```
<br />
<br />

Volcano plot:
```{r, message=FALSE}
input_df = DEGs
xlabel = "log2 FC Control vs Case"
ylabel = "-log10 adj_pvalue (FDR)"
par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$log2_FC, -log(input_df$p_adj,base=10),xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","red"), pch=20, frame.plot=TRUE, cex=0.8, main="Volcano plot") + abline(v=c(-1.5,1.5), col="black", lty=2) + abline(h=c(2), col="black", lty=2)
```
<br />
<br />

Heatmap: red labels are cases, green are controls. 
```{r}
cols = c(rep("green",50),rep("red",50)) 
pal = c("blue","white","red") 
pal = colorRampPalette(pal)(50)
heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$id[which(DEGs$class!="=")]),]),
        ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```
<br />
<br />

Gene set enrichment analysis of up-regulated and down-regulated genes separately using the Gene ontology: 
```{r, message=FALSE}
upDEGs = filtered_anno_df[rownames(DEGs %>% filter(class == "+")),]
downDEGs = filtered_anno_df[rownames(DEGs %>% filter(class == "-")),]

#GO - perform separately on BP and MF, then merge and sort by p.adjust
up_ego_BP = enrichGO(gene = upDEGs$symbol,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
up_ego_MF = enrichGO(gene = upDEGs$symbol,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
up_GO = rbind(up_ego_BP@result, up_ego_MF@result)
up_GO = up_GO[order(up_GO$p.adjust),]


down_ego_BP = enrichGO(gene = downDEGs$symbol,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
down_ego_MF = enrichGO(gene = downDEGs$symbol,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
down_GO = rbind(down_ego_BP@result, down_ego_MF@result)
down_GO = down_GO[order(down_GO$p.adjust),]

#KEGG
# get ENTREZ IDs
up_entrez_ids = getBM(attributes=c("ensembl_gene_id", "entrezgene_id"),
                filters=c("ensembl_gene_id"), 
                values=list(c(upDEGs$gene_id)),
                mart = ensembl)
# remove rows containing NA values 
up_entrez_ids = na.omit(up_entrez_ids)
# one Ensembl ID can be associated to multiple ENTREZ IDs - take only one association
up_entrez_ids = up_entrez_ids %>% distinct(ensembl_gene_id, .keep_all = TRUE)
# KEGG enrichment
up_ekegg = enrichKEGG(gene = up_entrez_ids$entrezgene_id,
                    organism = 'human',
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.05)
up_ekegg = up_ekegg@result
# sort
up_ekegg = up_ekegg[order(up_ekegg$p.adjust),]


# repeat for down-regulated
down_entrez_ids = getBM(attributes=c("ensembl_gene_id", "entrezgene_id"),
                filters=c("ensembl_gene_id"), 
                values=list(c(downDEGs$gene_id)),
                mart = ensembl)
down_entrez_ids = na.omit(down_entrez_ids)
down_entrez_ids = down_entrez_ids %>% distinct(ensembl_gene_id, .keep_all = TRUE)
down_ekegg = enrichKEGG(gene = down_entrez_ids$entrezgene_id,
                    organism = 'human',
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.05)
down_ekegg = down_ekegg@result
down_ekegg = down_ekegg[order(down_ekegg$p.adjust),]

```
<br />
<br />

Top 10 GO enrichment results (Molecular functions and Biological processes sub-ontologies combined) for up and down-regulated genes:
```{r}
up_GO[1:10, c(2,6)]
down_GO[1:10, c(2,6)]
```
<br />
<br />
Top 10 KEGG enrichment results for up and down-regulated genes:
```{r}
up_ekegg[1:10, c(2,6)]
down_ekegg[1:10, c(2,6)]
```
<br />
<br />

Using the list of up-regulated genes, none of the KEGG pathways was significantly enriched (adjusted p-value >=0.84). For this reason I visualized the **Estrogen signaling pathway** which had a significant adjusted p-value when using the list of down-regulated genes. 
PNG located in the folder
```{r, message=FALSE}
logFC = (DEGs %>% filter(class == "-"))$log2_FC
names(logFC) = down_entrez_ids$entrezgene_id
pathview(gene.data = logFC, 
         pathway.id = "hsa04915", 
         species = "human")
```
<br />
<br />

Motif enrichment analysis for **up-regulated** genes' upstream regions. Top 10 motif enrichment scores are shown below:
```{r}
upstream_seqs = biomaRt::getSequence(id = upDEGs$gene_id, type="ensembl_gene_id", seqType="gene_flank", upstream=500, mart=ensembl)
row.names(upstream_seqs) = upstream_seqs$ensembl_gene_id

data(PWMLogn.hg19.MotifDb.Hsap)
res = motifEnrichment(DNAStringSet(upstream_seqs$gene_flank),PWMLogn.hg19.MotifDb.Hsap, score = "affinity", verbose = FALSE)
report = groupReport(res)
report_df = as.data.frame(report)
report_df[1:10,c(2,4,5)]
```
<br />
<br />

Empirical distribution for the **CEBPB** transcription factor. 
```{r, message=FALSE}
tf = report_df$target[1]

tf_motif = subset(MotifDb, organism=='Hsapiens' & geneSymbol==tf)
PWM = toPWM(as.list(tf_motif))
ecdf = motifEcdf(PWM,organism = "hg19",quick=TRUE)
thresholds = lapply(ecdf,function(x) quantile(x,0.995))
```
<br />
<br />

Pattern-matching using all 19 PWMs of **CEBPB** binding sites in the flanking regions of up-regulated genes. 10 genes who's promoter regions have the highest count of scores above the threshold for any PWM are shown below:
``` {r}
scores = as.data.frame(motifScores(DNAStringSet(upstream_seqs$gene_flank),PWM,raw.score=FALSE,cutoff=thresholds, verbose = FALSE))
row.names(scores) = row.names(upstream_seqs)
scores$sum = c(apply(scores, 1, sum))
scores$symbol = lapply(upDEGs$symbol, function(x) toString(x))
scores = scores[order(-scores$sum),]

CEBPB_regulated_genes = subset(scores, scores$sum > 0)
CEBPB_regulated_genes[1:10,c(21,20)]

cat(sprintf("%s out of %s up-regulated genes have at least one region in their promoters with a binding score over the calculated threshold for ANY of the PWMs of CEBPB binding sites.", nrow(CEBPB_regulated_genes), nrow(scores)))
```

<br />
<br />

Upload files containing Ensembl IDs for up and down-regulated genes separately to String DB and download the generated PPI files. Mapping files from String DB were used to name the nodes. These files can be found in project directory.
``` {r}
# Export lists of ids of up and down regulated genes
#write.table(upDEGs$gene_id,file="upDEGs.txt" , quote=FALSE, row.names=FALSE, col.names=FALSE)
#write.table(downDEGs$gene_id,file="downDEGs.txt" , quote=FALSE, row.names=FALSE, col.names=FALSE)

# up-regulated PPI
up_links = read.delim("upDEGs_PPI.tsv")
up_mappings = read.csv("upDEGs_mapping.tsv", header=TRUE, sep="\t")
up_nodes = unique(up_mappings[,4])
up_net = graph_from_data_frame(d=up_links,vertices=up_nodes,directed=FALSE) 

up_c = components(up_net)
up_largest_c = induced_subgraph(up_net, which(up_c$membership == which.max(up_c$csize)) )


#down-regulated PPI
down_links = read.delim("downDEGs_PPI.tsv")
down_mappings = read.csv("downDEGs_mapping.tsv", header=TRUE, sep="\t")
down_nodes = unique(down_mappings[,4])
down_net = graph_from_data_frame(d=down_links, vertices=down_nodes, directed=FALSE)

down_c = components(down_net)
down_largest_c = induced_subgraph(down_net, which(down_c$membership == which.max(down_c$csize)) )
```
<br />
<br />

**Up-regulated** genes PPI:
```{r}
cat(sprintf("Using the up-regulated genes, the largest connected component contains %s out of %s elements.", max(up_c$csize), length(up_nodes) ))

plot(up_net, 
     edge.width=3,
     vertex.color="orange",
     vertex.size=10,
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.4,
     edge.curved=0.2, main="Complete up-regulated PPI") 

plot(up_largest_c,
     edge.width=3,
     vertex.color="green",
     vertex.size=10,
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.4,
     edge.curved=0.2, main="Largest connected component in up-regulated genes")
```
<br />
<br />

**Down-regulated** genes PPI:
```{r}
cat(sprintf("Using the down-regulated genes, the largest connected component contains %s out of %s elements.", max(down_c$csize), length(down_nodes) ))

plot(down_net, 
     edge.width=3,
     vertex.color="orange",
     vertex.size=10,
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.4,
     edge.curved=0.2, main="Complete down-regulated PPI") 

plot(down_largest_c,
     edge.width=3,
     vertex.color="green",
     vertex.size=10,
     vertex.frame.color="darkgray",
     vertex.label.color="black", 
     vertex.label.cex=0.4,
     edge.curved=0.2, main="Largest connected component in down-regulated genes")
```